pkg/controller/statusmanager/pod_status.go

```diff
@@ -72,37 +73,74 @@ func FindStatusCondition(conditions []corev1.NodeCondition, conditionType corev1
 	return nil
 }
 
-func SetNodeCondition(conditions *[]corev1.NodeCondition, newCondition corev1.NodeCondition) {
+func SetNodeCondition(conditions *[]corev1.NodeCondition, newCondition corev1.NodeCondition) (changed bool) {
+	changed = false
 	if conditions == nil {
 		conditions = &[]corev1.NodeCondition{}
 	}
 	existingCondition := FindStatusCondition(*conditions, newCondition.Type)
 	if existingCondition == nil {
 		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
 		*conditions = append(*conditions, newCondition)
-		return
+		return true
 	}
 
 	if existingCondition.Status != newCondition.Status {
 		existingCondition.Status = newCondition.Status
 		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		changed = true
 	}
 
 	existingCondition.Reason = newCondition.Reason
 	existingCondition.Message = newCondition.Message
+	return changed
 }
 
-func (status *StatusManager) setNodeNetworkUnavailable(nodeName string, ready bool, reason string, message string) {
+func (status *StatusManager) setNodeNetworkUnavailable(nodeName string, ready bool, reason string, message string, sharedInfo *sharedinfo.SharedInfo) {
+	if sharedInfo.LastNetworkAvailable == nil {
+		sharedInfo.LastNetworkAvailable = make(map[string]time.Time)
+	}
 	if ready == true {
 		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to false for node %s", nodeName))
 	} else {
 		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to true for node %s", nodeName))
+		delete(sharedInfo.LastNetworkAvailable, nodeName)
+	}
+	_, changed := status.combineNode(nodeName, ready, reason, message)
+	if !changed {
+		log.Info("Node condition is not changed")
+		return
+	}
+	if !ready {
+		status.updateNodeStatus(nodeName, ready, reason, message)
+		return
+	}
+	last := time.Now()
+	_, present := sharedInfo.LastNetworkAvailable[nodeName]
+	if !present {
+		sharedInfo.LastNetworkAvailable[nodeName] = last
+		log.Info(fmt.Sprintf("Last network unavailable time of node %s is set to %s", nodeName, last))
+
+		go func() {
+			time.Sleep(time.Duration(operatortypes.TimeBeforeRecoverNetwork) * time.Second)
+			if (sharedInfo.LastNetworkAvailable[nodeName]) != last {
+				log.Info(fmt.Sprintf("Last network unavailable time of node %s changed, expecting %v, got %v, goroutine exiting",
+					nodeName, last, sharedInfo.LastNetworkAvailable[nodeName]))
+				return
+			}
+			status.updateNodeStatus(nodeName, ready, reason, message)
+			return
+		}()
 	}
+}
+
+
+func (status *StatusManager) combineNode(nodeName string, ready bool, reason string, message string) (*corev1.Node, bool) {
 	node := &corev1.Node{}
 	err := status.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
```

Comment:
TODO: I think we should use informer to cache Node, instead of calling apiserver all the time

Reply:
make sense. put it in my queue.